src/asm/main.c

```diff
@@ -6,815 +6,96 @@
-		if ((instr.arg[i] & type) == 0)
-			ft_printf("invalid operand type: %s for instruction %s\n", ptr->operands[i], name);
-		ptr->op_types[i] = type;
-		if (type == T_REG)
-			ptr->size += 1;
-		else if ((type == T_IND) || ((type == T_DIR) && (instr.short_direct == 1)))
-			ptr->size += 2;
-		else if ((type == T_DIR) && (instr.short_direct == 0))
-			ptr->size += 4;
-		ptr->acb = update_acb(ptr->acb, i, type);
-		ptr->op_values[i] = ft_atoi(ptr->operands[i]);
-		i++;
+		ft_printf("{yellow}WARNING:{eoc} prog size (%d) ", a->prog_size);
+		ft_printf("exceeds CHAMP_MAX_SIZE (%d)\n", CHAMP_MAX_SIZE);
 	}
-	a->prog_size += ptr->size;
-
-	if (a->instruction == NULL)
-		a->instruction = ptr;
+	generate_output(a);
+	if (a->options == 1)
+		generate_hexdump(a);
 	else
 	{
-		tail = a->instruction;
-		while (tail->next != NULL)
-			tail = tail->next;
-		tail->next = ptr;
-	}
-	return ;
-
-	
-//
-//
-//
-// implement operands validation
-//
-//
-
-}
-
-void	process_instruction(t_asm *a)
-{
-	char	**split;
-	int		count;
-
-	split = ft_strsplit(a->tmp, ' ');
-	count = count_char_array(split);
-	if ((count == 1) || (count == 3))
-	{
-		if (!is_label(split[0]))
-			ft_printf("error: expecting label\n");
-		else
-			add_label_to_queue(a, split[0]);
+		of = prepare_output_filename(a->filename);
+		ft_printf("{green}SUCCESS:{eoc} writing output program to %s\n", of);
+		if ((put_file_contents(of, (unsigned char *)&a->header,
+				sizeof(t_header) + a->prog_size)) != 1)
+			(void)print_error();
+		free(of);
 	}
-	if ((count == 2) || (count == 3))
-	{
-		if (!is_instruction(split[count - 2]))
-			ft_printf("%s\nerror: expecting instruction\n", a->tmp);
-		else if (!is_operand(split[count - 1]))
-			ft_printf("%s\nerror: expecting operands\n", a->tmp);
-		else
-			add_instruction_to_queue(a, split[count - 2], split[count - 1]);
-	}
-	free(split);
-}
-
-void	parse_instructions(t_asm *a)
-{
-	char	*line;
-
-	while (get_next_line(a->fd, &line) > 0)
-	{
-		insert_into_char_array(a, line);
-		if (line != NULL)
-		{
-			string_prepare(line, a);
-			if (a->tmp[0] != 0)
-			{
-				process_whitespace(a->tmp);
-				process_instruction(a);
-//				ft_printf("%s\n", a->tmp);
-			}
-			free(a->tmp);
-			a->tmp = NULL;
-		}
-	}
-}
-
-
-int		is_direct_label(char *str)
-{
-	if ((str[0] == DIRECT_CHAR) && (str[1] == LABEL_CHAR))
-		return (1);
-	return (0);
 }
 
-int		does_label_exist(t_asm *a, char *str)
+int		init(char **av, int f_index, int opt)
 {
-	char			*label;
-	t_instruction	*ptr;
+	int		fd;
+	t_asm	*a;
 
-	label = str;
-	if (is_direct_label(str))
-		label = &str[2];
-	ptr = a->instruction;
-	while (ptr != NULL)
+	if ((fd = open(av[f_index], O_RDONLY)) != -1)
 	{
-		if (ptr->type == T_LAB)
-		{
-			if (ft_strcmp(ptr->label, label) == 0)
-				return (1);
-		}
-		ptr = ptr->next;
+		a = ft_memalloc(sizeof(t_asm));
+		a->options = opt;
+		a->fd = fd;
+		a->filename = av[f_index];
+		routine(a);
+		clean_up(a);
 	}
+	else
+		ft_printf("{red}ERROR:{eoc} %s\n", strerror(errno));
```

Comment:
You can flip the if statement into an early return to make this easier to read. Can also include the actual filename we failed to open in the error message.

```
if ((fd = open(...)) < 0)
{
  ft_perrorf(C_RED "Error:" C_END " open %s", av[f_index]);
  return 1;
}
/* ... rest of code follows ... */
```

Reply:
fixed